5 Techniques of Sampling Imbalanced Data

In [ ]:
import pandas as pd
import numpy as np
import imblearn

In [5]:
#Divide the dataset to 
df= pd.read_csv('Creditcard_data.csv')
#x contains all the features
x= df.drop('Class',axis=1)
#y contains the target
y = df['Class']

                                    SAMPLING I 

Random Under-Sampling with imblearn

In [ ]:
from imblearn.under_sampling import RandomUnderSampler

rus= RandomUnderSampler(random_state=42, replacement=True)
#x is the predictor, and y is the target variable
x_rus , y_rus = rus.fit_resample(x,y) 

In [9]:
x_rus

Time        V1        V2        V3        V4        V5        V6  \
103    69 -0.608027  0.306694  1.473004  1.191127  0.021265  0.888442   
438   315 -1.014693  0.444918  2.259005  0.336946 -0.492326 -0.649977   
273   194 -1.131517  1.016399  0.735810  1.166614  0.790236 -1.187196   
107    71  1.331897 -0.579962  0.422606 -0.897752 -0.746254 -0.056273   
72     47  1.197839  0.236828  0.509605  0.657659 -0.365488 -0.745101   
707   534 -1.393265 -1.095569  2.607795 -2.529060 -0.135265 -0.373239   
21     17  0.962496  0.328461 -0.171479  2.109204  1.129566  1.696038   
618   465  1.263769 -0.197755  0.326901  0.274104 -0.476462 -0.117241   
122    78 -0.622478  0.014881  1.052489 -1.644574 -0.538265 -0.835528   
1       0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361   
182   118  1.254914  0.350287  0.302488  0.693114 -0.371470 -1.070256   
244   164  0.073497  0.551033  0.451890  0.114964  0.822947  0.251480   
541   406 -2.312227  1.951992 -1.609851  3.997906 -0.522188 -1.426545   
623   472 -3.043541 -3.157307  1.088463  2.288644  1.359805 -1.064823   
639   484 -0.928088  0.398194  1.741131  0.182673  0.966387 -0.901004   
699   529 -2.000567 -2.495484  2.467149  1.140053  2.462010  0.594262   
717   539 -1.738582  0.052740  1.187057 -0.656652  0.920623 -0.291788   
766   574  1.257719  0.364739  0.306923  0.690638 -0.357792 -1.067481   

           V7        V8        V9  ...       V20       V21       V22  \
103  1.190491  0.042147 -0.446426  ...  0.374086  0.046506  0.165163   
438  0.542460 -0.201798  0.491987  ...  0.399388  0.090761  0.829358   
273  0.736469 -0.327992 -0.555549  ... -0.265971  0.009613  0.315739   
107 -0.750317  0.128484 -0.964682  ...  0.052489 -0.039070 -0.255174   
72   0.079497 -0.130536 -0.052263  ... -0.120981 -0.149876 -0.374211   
707  0.221446 -0.962847 -1.304498  ... -0.229820 -0.281365  0.563218   
21   0.107712  0.521502 -1.191311  ... -0.269321  0.143997  0.402492   
618 -0.334771  0.004088 -1.414382  ... -0.530571 -0.247243 -0.177197   
122  0.517234 -1.105962 -1.311741  ... -0.756577  0.418819 -0.762122   
1   -0.078803  0.085102 -0.255425  ... -0.069083 -0.225775 -0.638672   
182  0.086781 -0.202836  0.035154  ... -0.079756 -0.287592 -0.832682   
244  0.296319  0.139497 -0.123050  ... -0.043920 -0.128758 -0.381932   
541 -2.537387  1.391657 -2.770089  ...  0.126911  0.517232 -0.035049   
623  0.325574 -0.067794 -0.270953  ...  2.102339  0.661696  0.435477   
639  0.879016 -0.156590 -0.142117  ... -0.118283  0.066353  0.281378   
699 -2.110183  0.788347  0.958809  ...  0.620899  0.422452  1.195394   
717  0.269083  0.140631  0.023464  ...  0.002974 -0.179545 -0.192036   
766  0.094272 -0.210300  0.014455  ... -0.073779 -0.286856 -0.820658   

          V23       V24       V25       V26       V27       V28  Amount  
103  0.063974 -0.333792  0.131776 -0.217692 -0.082183 -0.116730  168.56  
438 -0.105669  1.005763 -0.098694  0.396781  0.307894  0.098746   69.99  
273  0.054210  0.294232  0.003877 -0.314159 -0.099512  0.122697    1.00  
107  0.109333 -0.328448  0.162254 -0.447276  0.020071  0.006231   14.48  
72   0.145516  0.414984  0.212156  0.181724 -0.016402  0.015994    3.63  
707 -0.129765  0.089264  0.317551 -0.288551 -0.863521 -0.519168   25.00  
21  -0.048508 -1.371866  0.390814  0.199964  0.016371 -0.014605   34.09  
618 -0.009729 -0.011026  0.491120 -0.315904  0.055419  0.013076    7.69  
122 -0.098940  0.502973 -0.655251  0.743419 -0.020568 -0.118060  104.94  
1    0.101288 -0.339846  0.167170  0.125895 -0.008983  0.014724    2.69  
182  0.128083  0.339427  0.215944  0.094704 -0.023354  0.030892    2.69  
244  0.151012 -1.363967 -1.389079  0.075412  0.231750  0.230171    0.99  
541 -0.465211  0.320198  0.044519  0.177840  0.261145 -0.143276    0.00  
623  1.375966 -0.293803  0.279798 -0.145362 -0.252773  0.035764  529.00  
639 -0.257966  0.385384  0.391117 -0.453853 -0.104448 -0.125765    1.00  
699  0.297836 -0.857105 -0.219322  0.861019 -0.124622 -0.171060 

In [10]:
y_rus

103    0
438    0
273    0
107    0
72     0
707    0
21     0
618    0
122    0
1      1
182    1
244    1
541    1
623    1
639    1
699    1
717    1
766    1
Name: Class, dtype: int64

                                SAMPLING II

Random Over-Sampling With imblearn

In [ ]:
from imblearn.over_sampling import RandomOverSampler

ros= RandomOverSampler(random_state=42)
x_ros, y_ros = ros.fit_resample(x,y)

In [24]:
x_ros

Time        V1        V2        V3        V4        V5        V6  \
0        0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388   
1        0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361   
2        1 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499   
3        1 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203   
4        2 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921   
...    ...       ...       ...       ...       ...       ...       ...   
1521   529 -2.000567 -2.495484  2.467149  1.140053  2.462010  0.594262   
1522   164  0.073497  0.551033  0.451890  0.114964  0.822947  0.251480   
1523   164  0.073497  0.551033  0.451890  0.114964  0.822947  0.251480   
1524   539 -1.738582  0.052740  1.187057 -0.656652  0.920623 -0.291788   
1525   472 -3.043541 -3.157307  1.088463  2.288644  1.359805 -1.064823   

            V7        V8        V9  ...       V20       V21       V22  \
0     0.239599  0.098698  0.363787  ...  0.251412 -0.018307  0.277838   
1    -0.078803  0.085102 -0.255425  ... -0.069083 -0.225775 -0.638672   
2     0.791461  0.247676 -1.514654  ...  0.524980  0.247998  0.771679   
3     0.237609  0.377436 -1.387024  ... -0.208038 -0.108300  0.005274   
4     0.592941 -0.270533  0.817739  ...  0.408542 -0.009431  0.798278   
...        ...       ...       ...  ...       ...       ...       ...   
1521 -2.110183  0.788347  0.958809  ...  0.620899  0.422452  1.195394   
1522  0.296319  0.139497 -0.123050  ... -0.043920 -0.128758 -0.381932   
1523  0.296319  0.139497 -0.123050  ... -0.043920 -0.128758 -0.381932   
1524  0.269083  0.140631  0.023464  ...  0.002974 -0.179545 -0.192036   
1525  0.325574 -0.067794 -0.270953  ...  2.102339  0.661696  0.435477   

           V23       V24       V25       V26       V27       V28  Amount  
0    -0.110474  0.066928  0.128539 -0.189115  0.133558 -0.021053  149.62  
1     0.101288 -0.339846  0.167170  0.125895 -0.008983  0.014724    2.69  
2     0.909412 -0.689281 -0.327642 -0.139097 -0.055353 -0.059752  378.66  
3    -0.190321 -1.175575  0.647376 -0.221929  0.062723  0.061458  123.50  
4    -0.137458  0.141267 -0.206010  0.502292  0.219422  0.215153   69.99  
...        ...       ...       ...       ...       ...       ...     ...  
1521  0.297836 -0.857105 -0.219322  0.861019 -0.124622 -0.171060    1.50  
1522  0.151012 -1.363967 -1.389079  0.075412  0.231750  0.230171    0.99  
1523  0.151012 -1.363967 -1.389079  0.075412  0.231750  0.230171    0.99  
1524 -0.261879 -0.237477 -0.335040  0.240323 -0.345129 -0.383563    1.00  
1525  1.375966 -0.293803  0.279798 -0.145362 -0.252773  0.035764  529.00  

[1526 rows x 30 columns]

In [25]:
y_ros

0       0
1       1
2       0
3       0
4       0
       ..
1521    1
1522    1
1523    1
1524    1
1525    1
Name: Class, Length: 1526, dtype: int64

                                SAMPLING III

Under-Sampling: Tomek Links

In [34]:
from imblearn.under_sampling import TomekLinks
from collections import Counter

tl= TomekLinks(sampling_strategy='majority')

x_tl , y_tl = tl.fit_resample(x,y)
print('Original dataset shape:', Counter(y))
print('Resampled dataset shape:', Counter(y_tl))


Original dataset shape: Counter({0: 763, 1: 9})
Resampled dataset shape: Counter({0: 756, 1: 9})


                                SAMPLING IV

Synthetic Minority Oversampling Technique (SMOTE)

In [35]:
from imblearn.over_sampling import SMOTE

smote= SMOTE()

x_smote , y_smote = smote.fit_resample(x,y)

print("Original dataset shape: ", Counter(y))
print("Resample dataset shape: ", Counter(y_smote))

Original dataset shape:  Counter({0: 763, 1: 9})
Resample dataset shape:  Counter({0: 763, 1: 763})


Smote has created 763 rows from 9 rows by using K-nearest neighbour algorithm on the rows classified as 1

                                SAMPLING V

NearMiss

In [36]:
from imblearn.under_sampling import NearMiss

nm = NearMiss()

x_nm, y_nm = nm.fit_resample(x,y)
print("Original Dataset Shape: ", Counter(y))
print("Resample Dataset Shape: ", Counter(y_nm))

Original Dataset Shape:  Counter({0: 763, 1: 9})
Resample Dataset Shape:  Counter({0: 9, 1: 9})


Applying 5 ML Models on these Samples to find Accuracy

In [44]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score, classification_report, f1_score
import warnings
warnings.filterwarnings('ignore')

models = {
    'Logistic Regression': LogisticRegression(max_iter=1000),
    'Decision Tree': DecisionTreeClassifier(random_state=42),
    'Random Forest': RandomForestClassifier(random_state=42),
    'SVM': SVC(random_state=42),
    'KNN': KNeighborsClassifier()
}

datasets = {
    'Random Under-Sampling': (x_rus, y_rus),
    'Random Over-Sampling': (x_ros, y_ros),
    'Tomek Links': (x_tl, y_tl),
    'SMOTE': (x_smote,y_smote),
    'NearMiss': (x_nm,y_nm)
}

results=[]

#Iterate over the different Sampling datasets to train all models
for sampling_name, (X_sampled, y_sampled) in datasets.items():
    X_train, X_test, y_train, y_test= train_test_split(X_sampled, y_sampled, test_size=0.3, random_state=42)
    #Apply models on each sample
    for model_name, model in models.items():
        #Train the model
        model.fit(X_train, y_train)
        #Predict
        y_pred= model.predict(X_test)
        #Calculate accuracy
        accuracy = accuracy_score(y_test,y_pred)
        f1= f1_score(y_test, y_pred)
        results.append({
            'Sampling Method': sampling_name,
            'Model':model_name,
            'Accuracy': round(accuracy, 4),
            'F1-Score': round(f1, 4)
        })

results_df= pd.DataFrame(results)
print(results_df)
results_df.to_csv('output.csv')


          Sampling Method                Model  Accuracy  F1-Score
0   Random Under-Sampling  Logistic Regression    0.8333    0.0000
1   Random Under-Sampling        Decision Tree    0.5000    0.4000
2   Random Under-Sampling        Random Forest    0.6667    0.5000
3   Random Under-Sampling                  SVM    0.1667    0.2857
4   Random Under-Sampling                  KNN    0.1667    0.2857
5    Random Over-Sampling  Logistic Regression    0.9192    0.9279
6    Random Over-Sampling        Decision Tree    0.9891    0.9896
7    Random Over-Sampling        Random Forest    1.0000    1.0000
8    Random Over-Sampling                  SVM    0.6856    0.6742
9    Random Over-Sampling                  KNN    0.9847    0.9855
10            Tomek Links  Logistic Regression    0.9870    0.0000
11            Tomek Links        Decision Tree    0.9826    0.0000
12            Tomek Links        Random Forest    0.9870    0.0000
13            Tomek Links                  SVM    0.9870    0.

In [45]:
pivot_table = results_df.pivot(index='Model', columns='Sampling Method', values='Accuracy')

# Reorder columns if needed
pivot_table = pivot_table[['Random Under-Sampling', 'Random Over-Sampling', 'Tomek Links', 'SMOTE', 'NearMiss']]

print(pivot_table)

# Save to CSV
pivot_table.to_csv('accuracy_comparison.csv')

Sampling Method      Random Under-Sampling  Random Over-Sampling  Tomek Links  \
Model                                                                           
Decision Tree                       0.5000                0.9891       0.9826   
KNN                                 0.1667                0.9847       0.9870   
Logistic Regression                 0.8333                0.9192       0.9870   
Random Forest                       0.6667                1.0000       0.9870   
SVM                                 0.1667                0.6856       0.9870   

Sampling Method       SMOTE  NearMiss  
Model                                  
Decision Tree        0.9672    0.1667  
KNN                  0.8362    0.8333  
Logistic Regression  0.9258    0.5000  
Random Forest        0.9934    0.1667  
SVM                  0.6900    0.1667  
